### PR TITLE
Cache `gem install fontist`, cache `~/.fontist`, and add semver support

### LINKED ISSUE #1: Add caching option
This action to setup Fontist takes around 1 minute.

We should add a `cache` option so that the setup is cached.

### LINKED ISSUE #7: Cache `~/.fontist`
offshot of #1 with explicit goal to cache the `~/.fontist` folder.

@ronaldtse what are the parts of ~/.fontist? should only certain subfolders in ~/.fontist be cached? all of it?

secondary question: should this be cached as part of $RUNNER_TOOL_CACHE or as part of the workflow cache? im thinking workflow cache since it's "dependency-like" ?

another secondary question: how should this be keyed so that it gets invalidated when the user changes something that affects the stuff that `fontist` installs? `manifest.yml`?

### LINKED ISSUE #8: Cache `gem install fontist`
ideally somehow want to avoid the expensive installation that happens. ideal world would be that fontist releases a single binary that we could just download and use but i dont think you can do that with ruby (?)

~~next best thing would be for fontist (or this action) to host a pre-compiled installdir result from `gem install fontist --install-dir /tarballme` for each Windows, macOS, and Linux.~~ not practical 

next best thing after that is to use $RUNNER_TOOL_CACHE and/or workflow cache to cache the result of `gem install fontist --install-dir /somewhere` so that the same user only gets hit with the 1 minute compile time ONCE and then just uses that cache over and over again

### LINKED ISSUE #9: Support semver specifiers in `fontist-version` field
make this work:

```yml
- uses: fontist/setup-fontist@v1
  with:
    # examples
    fontist-version: "1"
    fontist-version: "1.x"
    fontist-version: "^1.0.0"
    fontist-version: latest
```

### PR DESCRIPTION
**Pull request by jcbhmr**
This PR would:
1. use javascript instead of composite
2. use bun to bundle the code into a single file (dependencies included)
3. update github actions to have a build step upon releases
4. STILL REQUIRE THAT A GOOD VERSION OF RUBY IS USED see #5 -- idea to have own Ruby install location was abandoned; can be revisisted -- its not _that_ hard to do now that i did it once
5. first check $RUNNER_TOOL_CACHE for fontist
6. if `cache: true` and if no tool cache, try to restore from workflow cache
7. if still no fontist, resort to `gem install fontist` with custom `--install-dir` to avoid contaminating user's rubygems install location
8. create wrapper `fontist` and `fontist.cmd` scripts which proxy to the `--install-dir` location with some magic env vars set to load the library `.rb` files from the `--install-dir` instead of the global default spot
9. if `cache: true` and no cache content then try to stash in cache
10. hashFiles() of manifest.yml or manifest.yaml in the root folder. this is overridable by the user via cache-dependency-path just like setup-node
11. use hashFiles() as part of the cache key to try and restore ~/.fontist from cache; can fail if no current cache (which is OK)
12. run fontist update to refresh any outdated cache stuffs
13. POST RUN: save current filled ~/.fontist to cache with hashFiles() cach key from before; can fail if cache already exists (which is OK)

fix #9 
fix #8 
fix #7 
fix #1 
basically does a good amount of caching and adds semver version support

### CODE REVIEW

**Comment by jcbhmr** [comment]
ok so you CANNOT install ruby to a different location due to https://github.com/ruby/setup-ruby/blob/bd03e04863f52d169e18a2b190e8fa6b84938215/common.js#L261 BUT i think you can avoid adding it to the $PATH and just use the /opt/hostedtoolcache/Ruby/3.3.0/x64/bin/ruby binary directly

**Comment by jcbhmr** [comment]
actually i have to confirm that `gem install fontist` actually install things to the global RUNNER_TOOL_CACHE and not just to /var/somewhere. installing to the RUNNER_TOOL_CACHE is what we want for fontist. if that doesn't happen then there will still be 1 minute install times 🤔 🤔 😖

originally i wanted to do `gem install fontist --version ${version} --install-dir $RUNNER_TOOL_CACHE/install-dir` but that didn't work well https://github.com/fontist/setup/actions/runs/7673207894/job/20915332962?pr=6

any ideas ruby experts?

**Comment by jcbhmr** [comment]
so actually this DOES NOT make `gem install fontist` cached. the installation Ruby files go into /var/lib/gems/3.0.0/gems/ (i think) which is NOT part of $RUNNER_TOOL_CACHE and therefore not cached across runs. i retract any of my previous claims that "caching the installation is solved by #6" lol.

tl;dr: its installed to a place outside of caching ability and mixed with other deps (right now).

solution ideas:
- copy the installation files back and forth from the /var/... to the $RUNNER_TOOL_CACHE and back again to restore. problem is that the installations all have a version suffix like ...../package-1.2.3 which means that if this solution were used then anytime the deps changed of fontist this would break. it would also break across versions if the user specified `fontist-version: 1.10.0` or whatever

- create a wrapper script that sets GEM_HOME or whatevs https://stackoverflow.com/questions/16098757/specify-gem-installation-directory to the $RUNNER_TOOL_CACHE/fontist folder and use `gem install --install-dir` to place the installation ruby files there. this seems the most feasible.

- give up and deal with the 1 minute `gem install fontist` times

**Comment by jcbhmr** [comment]
i think the logic and results look good but id like someone else to review this ❤️ 

image


(yes this pr has changed names and purposes like 5 times lol 😆 )

**Comment by ronaldtse** [comment]
Thank you @jcbhmr for this!

@CAMOBAP @alexeymorozov could you please help review and merge this? Thanks!

File: .github/workflows/test-action.yml [hunk 1]
```diff
@@ -23,9 +23,13 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-latest, macos-latest, windows-latest]
+        # os: [ubuntu-latest, macos-latest, windows-latest]
```
Comment (CAMOBAP) [inline .github/workflows/test-action.yml:26]: Are there some issues on Windows discovered?

File: README.md [hunk 2]
```diff
@@ -31,19 +36,36 @@ jobs:
...
+- **`cache`:** Whether or not to use [@actions/cache](https://www.npmjs.com/package/@actions/cache) to cache things in the GitHub workflow cache. This is enabled by default.
+
+- **`cache-dependency-path`:** A multiline list of globs to use to derive the `~/.fontist` cache key. The default is `manifest.yml` and `manifest.yaml`. If no files are matched at runtime then the `~/.fontist` folder will not be cached.
 
-- **`fontist-version`:** An exact version of the fontist Ruby gem to install. Defaults to `1.19.0`.
+### Outputs
+
+- **`fontist-version`:** The version of Fontist that was installed. This will be something like `1.10.0` or similar.
+
+- **`cache-hit`:** Whether or not Fontist was restored from the runner's cache or download anew.
```
Comment (CAMOBAP) [inline README.md]:
```suggestion
- **`cache-hit`:** Whether or not Fontist was restored from the runner's cache or downloaded a new one.
```

File: README.md [hunk 3]
```diff
@@ -31,19 +36,36 @@ jobs:
       - run: fontist install "Fira Code"
 ```
 
-**⚠️ This action takes a while** (~1 minute). There's currently **no caching**. [Caching is planned for the future.](#TODO)
+💡 You can use `fontist manifest-install manifest.yml` to install fonts listed in a manifest file similar to `package.json`, `requirements.txt`, and `Cargo.toml`.
 
 ### Inputs
 
-- **`setup-ruby`:** Whether or not to run [ruby/setup-ruby](https://github.com/ruby/setup-ruby) or to assume that the user has supplied an appropriate global Ruby installation. Defaults to `true`. Set this to `false` to provide your own custom Ruby installation.
```
Comment (CAMOBAP) [inline README.md:38]: Probably we still has to note this implicit dependency on ruby/gem because nothing stops users from running int on some docker container without ruby

**Review by CAMOBAP — APPROVED** [review]
Good Job!

Few minor comments added

File: .github/workflows/test-action.yml (hunk 1 above)
Comment (ronaldtse) [reply .github/workflows/test-action.yml:26]: It's because `bun` is not supported on Windows.

File: README.md (hunk 2 above)
Comment (ronaldtse) [reply README.md]:
```suggestion
- **`cache-hit`:** Whether or not Fontist was restored from the runner's cache or newly downloaded.
```

Comment (ronaldtse) [reply README.md]: Easier to understand for folks not familiar with "anew".

File: README.md (hunk 3 above)
Comment (ronaldtse) [reply README.md:38]: I think the usage on docker containers will also apply to `node` since it's unclear whether a `node` environment would exist... Let's consider this a future enhancement.

**Review by ronaldtse — APPROVED** [review]

**Comment by ronaldtse** [comment]
Thank you @jcbhmr !

**Comment by ronaldtse** [comment]
@CAMOBAP we can probably start using this action for our Metanorma actions!
